### DISCUSSION TITLE
Learner app implementation

### DISCUSSION THREAD

**Discussion by sukhpreetssekhon** [Sunbird-RC/community#333] [Ideas & Enhancements]
A state education department plans to leverage RC to issue educational credentials (proof of enrolment, proof of scholarship / aid, proof of learning achievement) which need to be displayed in a learner app. Following flows are required in the app:

1. Learners can login to the app using a preferred auth system (with a OTP verification) and set username / password or mobile (OTP) for future logins
2. Learners can pull-in credentials issued against their common ID to their learner app
3. Learners can pull-in credentials from other compatible wallets (e.g., DigiLocker) by entering a common ID (e.g., roll number for 10th Board mark-sheet)
4. Learners can view any of the pulled-in credentials in the set visual template
5. Learners can download credentials as PDFs with QR code
6. Learners can share credentials as ‘links’ to verifiers via any medium (WA, SMS, Email)
7. Learners can hide credentials from their learner app and re-pull hidden credentials

Wireframes for the learner app are also linked here: https://whimsical.com/ulp-wallet-Ktz51kLahvhLQz31vuD2WA

Are there any implementations available in RC or in the larger open source ecosystem that can be contributed back to RC? Would the open source version(s) also come with a reference UI or only includes backend services? 

_Please note - We are using learner app as the terminology. In the past this has also been referred to as the learner's wallet._ 

@parthlawate @coolbung @suresh12 @chakshugautam @girirajdaga @surendrasinghs

**Comment by parthlawate** [comment]
Hi @sukhpreetsamagra 

In the eLocker implementation of Sunbird RC, some of this has been achieved. So that could be extended .. I am commenting inline to your requirements for the mapping. 

Elocker implementation details, link & repo https://docs.sunbirdrc.dev/sample-use-cases/digital-locker 

1. Learners can login to the app using a preferred auth system (with a OTP verification) and set username / password or mobile (OTP) for future logins
>>Mobile + Otp login already there. Uses Keycloak .. can be changed to username password as well. 

2. Learners can pull-in credentials issued against their common ID to their learner app
>> The Pull is implemented here based on integrated providers which is done as plugins. Right now there is a sample DIVOC and MOSIP Integration available 

4. Learners can pull-in credentials from other compatible wallets (e.g., DigiLocker) by entering a common ID (e.g., roll number for 10th Board mark-sheet)
>> Would be similar to above 

5. Learners can view any of the pulled-in credentials in the set visual template
>>Yes this is supported 

6. Learners can download credentials as PDFs with QR code
>>Yes this is supported 

7. Learners can share credentials as ‘links’ to verifiers via any medium (WA, SMS, Email)
>> Not Supported yet

8. Learners can hide credentials from their learner app and re-pull hidden credentials
>> Not Supported yet

Here is a quick video of the eLOcker Implementation to help you visualise. 

[elockerSunbird_rc.webm](https://user-images.githubusercontent.com/673176/216537152-9322b952-a394-4ab5-aef3-2642c12d1452.webm)

**Reply by suresh12** [reply]
> Hi @sukhpreetsamagra
> 
> In the eLocker implementation of Sunbird RC, some of this has been achieved. So that could be extended .. I am commenting inline to your requirements for the mapping.
> > @parthlawate eLocker Implementation can be extended for UI and additional use cases. 
> 
> Elocker implementation details, link & repo https://docs.sunbirdrc.dev/sample-use-cases/digital-locker
> 
> 1. Learners can login to the app using a preferred auth system (with a OTP verification) and set username / password or mobile (OTP) for future logins
> 
> > > Mobile + Otp login already there. Uses Keycloak .. can be changed to username password as well.
> 
> 2. Learners can pull-in credentials issued against their common ID to their learner app
> 
> > > The Pull is implemented here based on integrated providers which is done as plugins. Right now there is a sample DIVOC and MOSIP Integration available
> 
> 4. Learners can pull-in credentials from other compatible wallets (e.g., DigiLocker) by entering a common ID (e.g., roll number for 10th Board mark-sheet)
> 
> > > Would be similar to above
> 
> 5. Learners can view any of the pulled-in credentials in the set visual template
> 
> > > Yes this is supported
> 
> 6. Learners can download credentials as PDFs with QR code
> 
> > > Yes this is supported
> 
> 7. Learners can share credentials as ‘links’ to verifiers via any medium (WA, SMS, Email)
> 
> > > Not Supported yet
> 
> 8. Learners can hide credentials from their learner app and re-pull hidden credentials
> 
> > > Not Supported yet
> 
> Here is a quick video of the eLOcker Implementation to help you visualise.
> 
>  elockerSunbird_rc.webm

**Reply by sukhpreetssekhon** [reply]
@parthlawate & @suresh12 thanks for the updates. do share your thoughts on the way forward once you’ve discussed these in more detail today.

**Comment by sukhpreetssekhon** [comment]
@girirajdaga - Anything from Affinidi that can be contributed back along these lines to RC in the next few weeks?

**Reply by sukhpreetssekhon** [reply]
Do let us know as we would like to move ahead with an implementation plan today. @girirajdaga

**Comment by suresh12** [comment]
Thanks everyone. We have reviewed Sunbird E-Locker implementation https://docs.sunbirdrc.dev/sample-use-cases/digital-locker and decided to use it with minimal extensions to support above use cases.
